src/devices/sara_u280.c

```diff
@@ -0,0 +1,467 @@
+                                               msgs_len);
+        return is_rsp_ok(msgs, count);
+}
+
+static bool sara_u280_start_direct_mode(struct serial_buffer *sb,
+                                        const int socket_id)
+{
+        const char *msgs[1];
+        const size_t msgs_len = ARRAY_LEN(msgs);
+
+        serial_buffer_reset(sb);
+        serial_buffer_printf_append(sb, "AT+USODL=%d", socket_id);
+        const size_t count = cellular_exec_cmd(sb, MEDIUM_TIMEOUT, msgs,
+                                               msgs_len);
+        return is_rsp(msgs, count, "CONNECT");
+}
+
+static bool sara_u280_stop_direct_mode(struct serial_buffer *sb)
+{
+        /* Must delay min 2000ms before stopping direct mode */
+	delayMs(2100);
+
+        /*
+         * Using straight serial buffer logic here instead of
+         * the AT command system because we can get data intended for
+         * the JSON parser while we do this. Also because we don't
+         * want the \r at the end of the command.
+         */
+        serial_buffer_reset(sb);
+        serial_buffer_append(sb, "+++");
+        serial_buffer_tx(sb);
+        for (size_t events = 10; events; --events) {
+                serial_buffer_reset(sb);
+                if (serial_buffer_rx(sb, 1000) &&
+                    is_rsp_ok((const char**) &(sb->buffer), 1))
+                        return true;
+        }
+
+        return false;
+}
+
+static bool sara_u280_init(struct serial_buffer *sb,
+                           struct cellular_info *ci)
+{
+        /* NO-OP.  This hardware is better than sim900. */
+        return true;
+}
+
+static bool sara_u280_get_sim_info(struct serial_buffer *sb,
+                                   struct cellular_info *ci)
+{
+        return sara_u280_get_imei(sb, ci) &&
+                sara_u280_get_signal_strength(sb, ci) &&
+                sara_u280_get_subscriber_number(sb, ci);
+}
+
+static bool sara_u280_register_on_network(struct serial_buffer *sb,
+                                          struct cellular_info *ci)
+{
+        /* Check our status on the network */
+        for (size_t tries = 30; tries; --tries) {
+                sara_u280_get_net_reg_status(sb, ci);
+
+                switch(ci->net_status) {
+                case CELLULAR_NETWORK_DENIED:
+                case CELLULAR_NETWORK_REGISTERED:
+                        goto out;
+                default:
+                        break;
+                }
+
+                /* Wait before trying again */
+                delayMs(1000);
+        }
+
+out:
+        return CELLULAR_NETWORK_REGISTERED == ci->net_status;
+}
+
+static bool sara_u280_setup_pdp(struct serial_buffer *sb,
+                                struct cellular_info *ci,
+                                const CellularConfig *cc)
+{
+        /* Check GPRS attached */
+        bool gprs_attached;
+        for (size_t tries = 10; tries; --tries) {
```

Comment:
magic numbers applies througout.  Might be good for a config include or similar that applies to both modem technologies.